### ISSUE TITLE
[ARM32/ARM64] Introducing swap operation without temporary register

### ISSUE THREAD

**Issue by hqueue** [dotnet/coreclr#6273]
As you already know, there is a popular XOR swap which can exchange two values without temporary storage.

```
X := X XOR Y
Y := Y XOR X
X := X XOR Y
```
I just come up with some swap operations in ARM legacy backend which allocate a temporary register for swap operation and some swap operation in ARM64 codegen (some of them are NYI).

I'm going to exploit this XOR techqniue, especially for legacy backend to reduce register pressure where LSRA is not available. Of course, it's not always better when considering ILP, but removing temporary register may give us the freedom of implementation and design.

In addition, both A32 and A64 ISA from ARMv8 are lack of swap instruction such as `xchg `in x86 and x64.
FYI. ARMv8.1 introduces `CAS `and `SWP` related instruction but I think current coreclr only supports ARMv8 by now.

In short, I think XOR swap can be a candidate design choice for legacy backend of ARM32 to reduce register pressure where LSRA is not available.
For ARM64, this should be considered carefully, becuase of ILP in ARMv8 and LSRA of ryujit backend can address register pressure problem.

I will make a PR for simple case.
What do you think of this idea?

**Comment by mikedn** [comment]
Note that XOR swap won't work on object references and managed pointers, the intermediary X it produces can't be tracked by GC.

**Comment by hseok-oh** [comment]
@mikedn XOR swap won't work on object references and managed pointers in CIL(ex. C# code) level. @hqueue 's idea is in machine code (generated by JIT compiler) level.

**Comment by mikedn** [comment]
> XOR swap won't work on object references and managed pointers in CIL(ex. C# code) level. @hqueue 's idea is in machine code (generated by JIT compiler) level.

The GC doesn't care about CIL, it cares about what the generated machine code does. If you have 2 object references in 2 registers and want to swap them it won't work with a XOR swap.

**Comment by mikedn** [comment]
Well, maybe it will work if you prevent GC from running during the XOR swap.

**Comment by mikedn** [comment]
> How does GC know which registers or memory(stack, heap) location contain address at runtime ? (conservative GC, precise GC?) Is there any information maintained and JIT should update the information?

Yes, the JIT emits information for the GC, see gcinfo.cpp in the JIT source directory. There's a bit of info in the RyuJIT doc's [GC Info](https://github.com/dotnet/coreclr/blob/master/Documentation/botr/ryujit-overview.md#gc-info) section.

**Comment by gkhanna79** [comment]
CC @sdmaclea

**Comment by hqueue** [comment]
I'm going to close this issue if there is no more feedback, since PR was already merged long time before and there is no change regarding GC traking for references in this case.

**Comment by sdmaclea** [comment]
@gkhanna79 Thanks.  I looked at patch #6274 which implemented this feature for ARM32.  Looks like it only applies to Arm32 on legacy code gen.  I was assume you mentioned me to look at the GC issues this might have introduced on arm64.  However I see that you added the enhancement tag above.  Should an issue like this be opened for RyuJit enhancement for arm & arm64

**Comment by gkhanna79** [comment]
Yes, I CCed you @sdmaclea incase there needs to be similar work that needs to be done for Arm64 Linux and if so, do file appropriate issues for it.

**Comment by kasper3** [comment]
> I'm going to close this issue if there is no more feedback, since PR was already merged long time before

Yes please!

**Comment by hqueue** [comment]
@BruceForstall Can we close this or leave it open for RyuJIT enhancement ?

**Comment by BruceForstall** [comment]
@hqueue seems reasonable to leave open to track for RyuJIT enhancement.

**Comment by BruceForstall** [comment]
LSRA generates `GT_SWAP` nodes to generate register swaps, if necessary, during block resolution, for integer registers only, for x86/x64 only. See `resolveEdge()` and `insertSwap()`. Otherwise, it spills to stack then reloads from stack. It seems we could extend `GT_SWAP` to use the 3-instruction XOR code on arm32/arm64 for non-GC-ref pointers, using the model previously implemented for legacy arm32 back-end: https://github.com/dotnet/coreclr/pull/6274.

Swaps like this are expected to be relatively rare, so it's not clear how much this would buy us.

There is a bunch of code in codegenarm64.cpp for `genCodeForSwap()` which is a copy of the xarch code, with an NYI. So we should either implement it or decide not to and remove this code.

**Comment by sdmaclea** [comment]
LSE atomic also added a `swp` instruction to arm64.  So at least on some newer arm64 systems this will become available.
